### DISCUSSION TITLE
Mouse Input Lag

### DISCUSSION THREAD

**Discussion by retropcfan** [86Box/86Box#3779] [Q&A]
### What happened?

Hello,

I have used 86Box and PCem for quite some time now and there seems to be an issue with 86Box. Not many people seem to talk about it but it's something I really notice. The mouse input lag in 86Box seems to be much worse than it is in PCem. 

I've tried pretty much anything you cold think of:

-all versions of 86Box, including the latest builds (both 32-Nit and 64-Bit
-different host pcs
-different host pc mice
-different emulated mice in 86Box (serial, PS/2)
-different operating systems in 86Box
-emulation speed is always at 100%
-and so on...

I always compare it directly to PCem and real hardware. For example I use the exact same configuration as PCem and the same harddrive wich I just copied (I also made new virtual machines from the ground up in 86Box).

To compare the mouse behavior a little better I used Microsoft Paint in Windows and draw cricles. In PCem (and real hardware ofcourse) the circles are perfectly smooth, just like you expect them to. In 86Box however, the circles are weird, they have straight lines in them when you move the mouse a little faster, almost like it always skips a few frames or refreshes.

Thing is, if it would just be in paint than I would really care about it. But you do feel it as well when using Windows in 86Box. The mouse is all laggy. To me this is so annoying that I still use PCem instead of 86Box, even though I otherwise prefer 86Box over PCem. 

I can't be the only one who notices this? 

If you don't believe me just try it out for yourself. Open 86Box and PCem at the same time and go back and forth. You will see how much smoother PCem is.

### Operating system

Windows 10 and 11

### CPU

i9 10850K and others

### 86Box version

5083

### Build architecture

Windows - x64 (64-bit)

### Build type

- [ ] New recompiler
- [ ] Debug build

### Download source

Official website (Jenkins, GitHub)

**Comment by OBattler** [comment]
What emulated CPU speed? From what I know, PCem does rather simplistic mouse emulation - update after every CPU frame and not emulating actual update rates. So the slower the emulated CPU, the smaller the CPU frame, the more frequent the mouse updates, the smoother the mouse. 86Box, on the other hand, actually emulates the mouse update rates so you need some tool (ps2rate or Microsoft IntelliPoint) to raise from the default (usually decided by the combination OS and BIOS) that tens to be quite low (my logging sessions have shown rates as low as 40-something Hz).

**Comment by OBattler** [comment]
As for the serial mouse, nominally, we do the same thing as PCem - updating after every CPU frame. But there's a catch - PCem blasts the serial data bytes at the guest instantly while we emulate the actual serial transmission speed, and the Microsoft serial mouse uses a speed of 1200 bps. That's going to limit the amount of packets per second by a lot, the Logitech C7 serial mouse technical reference actually warns about that.

So PCem achieves its universal smoothness (that also isn't really smooth on faster emulated CPU speeds, anyway) by cutting down on accuracy. 86Box emulates the mice more accurately but that's going to mean lower report rates in some case, much as would occur on real hardware. In some cases, eg. for PS/2 mice, there are remedies (as said above, ps2rate or Microsoft IntelliPoint) but in other cases (eg. serial mice), there is unfortunately none.

But you can set the MS/Logitech bus mouse (not InPort) with the rate set to continuous to have something more like PCem.

**Comment by cold-brewed** [comment]
I'm curious as to which guest operating systems you've tried? I have noticed windows 9x leaves the ps/2 rate at whatever is set by the bios (which can be quite slow). 86Box leaves this as-is for accuracy purposes. 

There's a program called `ps2rate` that you can run in 98 to manually set the ps2 rate (also works on real hardware) and you should notice a sharp difference once you manually set the ps/2 rate up higher.

Another observation: It seems that NT 4 and greater *will* in fact set the ps/2 rate up much higher by default, even when the bios sets it to a low initial value. In my testing NT4, 5, 5.1 all have extremely smooth mouse performance in 86Box.

I have also observed the same effect in other linux distributions (excluding 90s distros) such as early versions of ubuntu. Although overall the system will run slow on a pentium II, the mouse is extremely smooth because the ps/2 rate gets set up to the maximum rate by the OS.

Do you have a system that you could share for us to check? One where you're having sluggish mouse performance.

I'd also recommend trying NT 4 with a ps/2 mouse and seeing if the results are better.

**Comment by todopoderoso1965** [comment]
Agree. 86Box has a serious mouse lag. Compare Rise of the Triad on 86Box and DosBox. Difference is astonishing.

**Reply by OBattler** [reply]
I reiterate coldbrewed's question above:
> Do you have a system that you could share for us to check? One where you're having sluggish mouse performance.

As for this:

> Agree. 86Box has a serious mouse lag. Compare Rise of the Triad on 86Box and DosBox. Difference is astonishing.

I'd like to see a comparison with real period hardware, not with DOSBox. Given how much faster disk (including floppy and CD-ROM) access is on DOSBox compared to real old hardware and 86Box, I'm inclined to believe their mouse performance is also unrealistically fast.

**Comment by RicardoRamosWorks** [comment]
I believe I have the same problem. I created a 1:1 replica of a real computer on the 86box, and the mouse has a delay, if I move quickly from side to side, and suddenly stop, I can see the cursor "finishing" the movement, a type of input lag of half a second.

My machine is emulating at 100% all the time, even when opening Windows Explorer. PCEm also has a delay, but it is much smaller (and more tolerable)

**Reply by OBattler** [reply]
Input lag of half a second? I've never experienced anything like that. What emulated graphics card are you using? Perhaps whichever you're using, has incorrect timings that make it laggier than real hardware.

**Comment by igghcz** [comment]
Perhaps there are two problems meeting here. Mouse responsiveness and redraw speed. I've tried DosBox, PCem and 86Box and have real hardware in mind. I am comparing in pure DOS in text mode. FreeDos or Win98 Dos, Volkov Commander or mouse test utilities (gtest.exe, mousecc.exe), gmouse or ctmouse, logitech mouse driver. Different host PCs, different graphics cards, Windows 10. Guest was 486 with Intel chipset, Intel DX-25, 8 MB, T8900D. Changing the processor for a faster one, or changing cycles had no significant effect.

The response and redraw speed on real hardware of this configuration in text mode was snappy. I estimate about 10 observable mouse cursor positions with very fast mouse circling. Similar to the 60 Hz LCD in Windows today. 

DosBox response and redraw are very snappy, response maybe a bit slower than real HW. (At about 4000-12000 cycles.) With fast horizntal mouse oscillation from one side of the screen to the other, certainly 8 cursor positions are displayed, including the extreme positions. (But you can see in the video that the cursor doesn't always reach the other side of the screen.)

PCem is a bit slower both on response and on the speed of redrawing. With an emulated serial mouse, I observe 6 displayed cursor positions while oscillating. (ctmouse /R1 /M)

86Box 4.2 with serial mouse response and redraw is slower than PCem. Only 4 displayed cursor positions when mouse oscillates. When using the emulation recommended by Logitech/MS Bus Mouse the response really improves and is about comparable to PCem. Hurray. But the redraw speed is still probably only about 4 cursor positions. However, when switching Render from QT SW, to QT OpenGL or OpenGL 3.0 Core, the redraw is even slower and it is downright easy to see the cursor arriving on the other side of the screen with a delay. Still, the cursor position only flashes in about 4 places on the screen.

With OpenGL 3.0 Core, I've noticed several times that the rectangular text mode cursor briefly turns into an arrow - the windows cursor.

[Videos.zip](https://github.com/user-attachments/files/16625900/Videos.zip)

**Reply by OBattler** [reply]
Are you comparing this to host hardware from the same period or to modern host hardware? Comparing mouse on emulated ancient hardware to modern host hardware is pointless. Modern mice have a vastly higher report rate than older mice. Thu bus also plays a part - serial is slow and is going to limit how many reports can be sent per second. For PS/2 - the default rate set by most older BIOS'es is low, and you need something like ps2rate.exe or Microsoft IntellISense or what it's called to have a proper rate.

**Reply by igghcz** [reply]
In a previous post, I compared the response of an emulated 486 with an emulated serial mouse.
The host PCs were similar configurations, all about 10 years old, and behaved similarly:
1. All in One, Intel Core i5-2400S 2.5GHz, AMD graphics.
2. NB with integrated Intel.
3. NB with integrated Intel and Nvidia graphics. 

On these host PCs I compared the response of the DosBox, PCem and 86Box against how I remember the response on the 486 from the 90s. I compared in emulated DOS text mode.

I also tried various Renders on the host PCs.
While in DosBox I didn't notice much difference (between renders) in 86Box the results were much better with QT SW than with GL modes - ok, old host HW. 

In 86Box the emulated serial mouse had poor response compared to how I remembered. The Logitech/MS Bus Mouse was a response close to memory.

Over the weekend I installed the 86Box on Nobara Linux with a 3.6 GHz i7 and GTX1070. In the 86Box, the emulated serial mouse had poor response compared to how I remembered. The Logitech/MS Bus Mouse was a response close to memory. GL renders were usable, the Vulkan option was not available. I didn't investigate thoroughly.

Today I fired up a real old Pentium 166MMX with DOS and tested the response of a real serial mouse. With ball. I'm attaching a video shot at 60 fps. The LCD is very old from 2007. But even so, you can see the cursor response on the LCD compared to the mouse movement only 3-4 frames behind the mouse movement. That's 0.06 seconds. And this lag is not getting any bigger. The cursor is a bit delayed, but perfectly synchronized with the mouse movement. The cursor doesn't end the movement with more lag. 
If it would help, I will make the same video with the emulated environment.

> P.S. Are you against having two modes for mouse emulation?

I hope you will be able to do this. Thank you.

[Real_P166_SerialMouse1.zip](https://github.com/user-attachments/files/16664413/Real_P166_SerialMouse1.zip)
[Real_P166_SerialMouse2.zip](https://github.com/user-attachments/files/16664439/Real_P166_SerialMouse2.zip)
[Real_P166_SerialMouse3.zip](https://github.com/user-attachments/files/16664446/Real_P166_SerialMouse3.zip)

**Comment by Roundabout1776** [comment]
Currently playing with 86Box under Win11 and Fedora 40 KDE/Wayland. I use ps2rate for 9x OSes and my panel is 240Hz, these seem to make mouse issues very apparent.
While I'm getting various issues with QT/GL modes on Fedora, Vulkan works reliably AND has less mouse lag than Windows. Could someone having issues under Windows try Linux and confirm that? That's important since apparently lots of work went into accurate mouse emulation and having it work noticeably different across various OSes and renderers just destroys it.

**Reply by OBattler** [reply]
The guest is not even close to 240 Hz. Also, we support Vulkan on Windows as well, have you tried that?

I also have no way of knowing the difference, given that I'm on a 60 Hz monitor.

I also, again, recommend observing real period hardware and seeing how mouse movement is there, as opposed to comparing it to a modern host machine.

**Reply by Roundabout1776** [reply]
>The guest is not even close to 240 Hz.

Yes but setting ps2rate from 80Hz to 200Hz gets me surprisingly close to my host.
I don't know any such tools for NT 4.0 and while this OS has better mouse than 9x out-of-the-box, it's probably set up to 120Hz or something, so a bit worse than 200Hz 9x.

>Also, we support Vulkan on Windows as well, have you tried that?

Yeah, Linux is just smoother for me. However I don't insist that it's universally the case hence asking people having similar issues to try and confirm it. At the end of the day it could be host mouse driver/setup issue.

>I also, again, recommend observing real period hardware and seeing how mouse movement is there, as opposed to comparing it to a modern host machine.

Unfortunately I have no access to real hardware.

P.S. Are you against having two modes for mouse emulation? So people interested in accuracy could use current implementation while others would prefer performance/latency.

**Reply by OBattler** [reply]
> P.S. Are you against having two modes for mouse emulation? So people interested in accuracy could use current implementation while others would prefer performance/latency.

I would have no problem with that, the problems are more on how to implement that - I'd basically have to directly send the host delta's to the guest, but that would entail either writing to the guest directly from the UI thread where input is grabbed, which would mean cross-thread accesses, which I actually tried during the rework in 4.0 and it resulted in the guest mouse actually being *slower* due to the PIC emulation (for the IRQ's) not being thread-safe, or queue the individual delta's, but then you'd have a delay, anyway.

But I have been thinking of how to solve the problem, I just haven't found a solution yet.

**Comment by Pellzi** [comment]
I have had nearly similar problems with the mouse lag. I installed Win 98.
I solved this problem by myself by doing this:

Change the Mouse input on x86Box Settings to Standard PS/2 Mouse.
Download the Program above from: https://www.majorgeeks.com/files/details/ps2rate.html
Unzipped and installed it into my virtual Win 98. I putted the included vxd Driver into Windows/System.
Then i run this program, and set the Default to Win Default. Pressed OK .
And this was it. No more Mouse lags in my virtual Win 98.

**Reply by OBattler** [reply]
Yes, unfortunately, this is needed because most BIOS'es and drivers initialize the PS/2 mouse to a very slow rate, so you need something like PS2Rate to increase it. On Windows 3.1x, I found that installing Microsoft IntelliPoint tends to help.

**Comment by Pellzi** [comment]
From:
https://github.com/JHRobotics/softgpu?tab=readme-ov-file#turn-on-dma

Turn on DMA
Windows by default using interrupts to access HDD and CD drive. This is especially slow in the HW accelerated virtual machines because every interrupt will stop executing visualisation engine and hand over access to hypervisor to solve it and after that it'll be need to restart visualisation engine again. Time consumed by interrupts depends on visualization technology for example in QEMU + kvm is HDD access very slow equally VirtualBox + Hyper-V and combination QEMU + Hyper-V is downright painful. Fortunately there is DMA transfer of whole memory block instead of individual bytes. Unfortunately you have to enable it manually.

You can turn it on in Device Manager on HDD properties enable DMA checkbox. Do it the same for CD driver and reboot VM for applying changes.
![dma](https://github.com/user-attachments/assets/b65a4e79-84be-427d-ad38-f4ed23e5616b)


This solves quite a bit the Problem of the Mouse lag. Together with the ps2rate.exe, the mouse behaveour is like it should.

**Reply by OBattler** [reply]
This is actually a very good observation and something that completely escaped my mind for some reason - mouse lag caused by the CPU not responding to interrupts fast enough due to it being busy with hard disk interrupts.

But to clarify - on 86Box, the CPU is entirely emulated, so an interrupt will make the CPU jump to the interrupt's vector, just like on real hardware. But the more devices are requesting interrupts, the longer it's going to take to process all of them, so the workaround still applies.

**Reply by 2mg** [reply]
I have to chime in with the same issues - tho note that I'm using one of the latest nightly builds (not dev/debug versions).

Latest successful artifacts - Old recompiler - Windows x64 - 86Box-Windows-64-b7043.zip > note that there is no 32bit version which is recommended in documentation for nightlies.

I've tried every option for Microsoft/Logitech Bus mouse, and PS/2 with PS2rate.

Mouse never feels snappy even in desktop, and I've tried some DDraw/GDI low requirement games where the mouse is still laggy.

Guest is PII 200Mhz, Voodoo3 stock settings, SB16, 1xHDD, 2xCD, DMA enabled, W98SE.

Host is Ryzen 5800x.

PS: which Intellipoint software can adjust polling rate?

**Comment by PoloniumRain** [comment]
This problem also exists with faster machines like a PII 450MHz, AMD K6-III 400MHz and Cyrix III 733MHz on a 9950X3D at 100% emulation speed. I'm using a build with the new recompile from 2 days ago, but mouse input latency has always been an issue.

But i just wanted to add that the emulated machines speed, and the host CPU or GPU speed, has no affect at all on the mouse input latency. It's poor on both Win10 and 11 as well. Noticeable mouse input latency also exists with PCem and PCBox. It's been a while since i used PCem, but i remember it maybe having slightly quicker input latency, although it was still noticeable.

Enabling DMA has no effect.

PS2rate.exe helps in the sense that it makes the mouse cursor movement look smoother on the Win98 desktop because the cursor position is visibly updated more often when PS2rate is set to 200Hz, and camera movement in FPS games is also clearly smoother (Quake 3 is one of **many** examples, where as previously i'd have to enable mouse smoothing in the games Options menu). But that's all PS2rate does. It has little to no effect on input latency.

If i had to guess, i'd say the mouse input delay is around 60 - 80ms, which is 4 - 5 frames at 60fps. Vulkan adds another 1 - 2 frames. Vulkan definitely has more latency than OpenGL and the Software renderer, at least on a host running Win10/11.

I also tested keyboard input latency with a high speed camera and compared it to DOSBox Pure. It was still hard to tell but it looks like there's a slightly longer input delay, but it's nowhere near as noticeable as the mouse delay.

And i don't have real hardware to compare it against.

**Reply by 2mg** [reply]
The fact that PCem has less input lag (as explained here, I presume that's really the differentiating factor and not something else hiding in the bush) at least means 86box could potentially look into other ways of handling the mouse.
OTOH PCem also has it too, so these two emulators share a common ancestor so to speak, I presume.

A patch/hack/plugin that fixes this, even if it's not a purist's approach, would probably be welcomed by everyone, as it would be on a per persons' case basis if they want it or not.

If real input lag is your mouse lagging behind your mouse movement, the "polling lag" is you visually waiting for the new mouse position, especially in desktop. Sure, 50Hz is _enough_, but it's felt in a way that detracts from the experience.
That's why PS2rate at 200Hz feels smoother, so it's definitely a part of the "mouse lag experience".

Vulkan's lag might be dependent on it's DXGI/Native method and/or how frames are displayed/Vsync'd by renderer and W10/11?

But what if the mouse lag difference between these emus isn't mouse lag, but render lag (maybe Vsync)?
As in, one has bigger render/display lag than the other, and I'm not talking about emulation speed.

What we need here is someone with any old working PC to compare (especially if they have few different mice).

**Reply by OBattler** [reply]
Try on the latest built with the OpenGL-Core 3.0 renderer, Vsync off, and sync to video on.

**Reply by 2mg** [reply]
Just replying that I had no time, sorry, but I'm watching this thread develop nonetheless!

**Comment by Cacodemon345** [comment]
This shouldn't be an issue anymore with the new "smaller frames" feature that allows for more frequent input polling.

**Comment by PoloniumRain** [comment]
It still has the same problem, but seems slightly improved.

I've found that Intel has the least amount of mouse input latency. AMD and Cyrix are worse, at least with the setups i'm using. Here's my machines:

AMD:
Super Socket 7
FIC VA 503+
K6-III 400MHz
Falcon Northwest Voodoo 3 SE
Sound Blaster PCI 128
PS/2 mouse
Win98 SE

Cyrix:
Socket 370
Asus CUV4X-LS
733MHz
Everything else is identical to AMD.

With an Intel PII 450MHz on a Slot 1 Gigabyte GA-686BX motherboard, and everything else exactly the same as the AMD machine, the mouse input **definitely** has lower latency. Intel is now good overall and i can play games without being annoyed by mouse input lag.

Could this be something to do with how the AMD and Cyrix motherboard BIOS handle mouse input? I don't have real hardware to test this.

**Reply by OBattler** [reply]
What emulated mouse? If it's PS/2, then we emulate the correct reporting rate, so you need something like ps2rate to increase that.

And even with the Logitech serial and bus mice where we support constant reports, those are going to only update every CPU frame, which means around every ms using smaller frames and every 10 ms using larger frames.

**Comment by PoloniumRain** [comment]
Ok i've stopped being lazy and tested the input latency in Quake 2 with a high speed camera at 960fps:


https://github.com/user-attachments/assets/851d9349-8747-481c-8598-d10b37cde2df


Intel, AMD and Cyrix are all very similar with around 1.5 frames of delay, which is good and better than before. @OBattler they all use PS/2 mice with ps2rate set to 200Hz.

I also tested mouse button presses and they were all very similar again. This time they each had around 3 frames of delay. Although maybe it took longer because Quake 2 simply takes longer to play the weapon firing animation, because it feels similar to natively running Q2.

However on the Windows 98 desktop, the AMD and Cyrix definitely have more input latency than the Intel. Mouse movement is around 3 frames of delay. The Win98 mouse settings are all identical so i don't know how this happens, but i'm not really bothered, it's just the desktop and still very much usable. The Intel also feels slightly faster in Quake 1, but only by an absolutely tiny amount. 

Overall games now feel responsive with this latest build, very close to native/host 60Hz, at least with the five Win98 games that i've tested so far (Quake 1, 2, 3, Hexen II, and Star Trek Voyager: Elite Force).

**Comment by Cacodemon345** [comment]
@retropcfan @2mg @PoloniumRain 

https://github.com/Cacodemon345/86Box/actions/runs/16328438504/artifacts/3548613194

Dev NDR MINGW64 build.

I'm testing a new method adapted from SDL3 into 86Box for mouse polling. This should hopefully lead to lessened mouse input lag/latency. Please test.

If it works out it'll be merged upstream. In my testing it was getting mouse events under 1-2 ms with my Lenovo IdeaPad Gaming M100 mouse.

**Comment by PoloniumRain** [comment]
@Cacodemon345 Everything is the same as my last post when using a high speed camera in Quake 2 again.

Also just tested more Win98 games (but just playing them, not using the camera):

Forsaken
Half-Life
Kingpin
MDK 2
Soldier of Fortune
Sin
Star Trek: Deep Space Nine: The Fallen
Turok
Unreal
Unreal Tournament

I didn't find any issues and all games feel perfectly responsive. Infact it's now pretty hard for me to feel the difference between Quake 2 at 60Hz in 86Box and Quake 2 remastered running natively on Win11 at 120Hz with v-sync disabled. It's way better!
